jax/_src/scipy/stats/_core.py

```diff
@@ -0,0 +1,151 @@
+    Parameters
+    ----------
+    a : array_like
+        n-dimensional array of which to find mode(s).
+    axis : int or None, optional
+        Axis along which to operate. Default is 0. If None, compute over
+        the whole array `a`.
+    nan_policy : {'propagate', 'raise', 'omit'}, optional
+        Defines how to handle when input contains nan.
+        The following options are available (default is 'propagate'):
+          * 'propagate': returns nan
+          * 'raise': throws an error
+          * 'omit': performs the calculations ignoring nan values
+    Returns
+    -------
+    mode : ndarray
+        Array of modal values.
+    count : ndarray
+        Array of counts for each mode.
+    Examples
+    --------
+    >>> a = jnp.array([[6, 8, 3, 0],
+    ...               [3, 2, 1, 7],
+    ...               [8, 1, 8, 4],
+    ...               [5, 3, 0, 5],
+    ...               [4, 7, 5, 9]])
+    >>> from scipy import stats
+    >>> stats.mode(a)
+    ModeResult(mode=array([[3, 1, 0, 0]]), count=array([[1, 1, 1, 1]]))
+    To get mode of whole array, specify ``axis=None``:
+    >>> stats.mode(a, axis=None)
+    ModeResult(mode=array([3]), count=array([3]))
+    """
+
+    a, axis = _chk_asarray(a, axis)
+    if a.size == 0:
+        return ModeResult(jnp.array([]), jnp.array([]))
+
+    contains_nan, nan_policy = _contains_nan(a, nan_policy)
+
+    if contains_nan and nan_policy == 'omit':
+        a = ma.masked_invalid(a)
+        return mstats_basic.mode(a, axis)
+
+    if a.dtype == object and jnp.nan in set(a.ravel()):
+        # Fall back to a slower method since jnp.unique does not work with NaN
+        scores = set(jnp.ravel(a))  # get ALL unique values
+        testshape = list(a.shape)
+        testshape[axis] = 1
+        oldmostfreq = jnp.zeros(testshape, dtype=a.dtype)
+        oldcounts = jnp.zeros(testshape, dtype=int)
+
+        for score in scores:
+            template = (a == score)
+            counts = jnp.sum(template, axis, keepdims=True)
+            mostfrequent = jnp.where(counts > oldcounts, score, oldmostfreq)
+            oldcounts = jnp.maximum(counts, oldcounts)
+            oldmostfreq = mostfrequent
+
+        return ModeResult(mostfrequent, oldcounts)
+
+    def _mode1D(a):
+        vals, cnts = jnp.unique(a, return_counts=True)
+        return vals[cnts.argmax()], cnts.max()
+
+    # jnp.apply_along_axis will convert the _mode1D tuples to a numpy array,
+    # casting types in the process.
+    # This recreates the results without that issue
+    # View of a, rotated so the requested axis is last
+    in_dims = list(range(a.ndim))
+    a_view = jnp.transpose(a, in_dims[:axis] + in_dims[axis+1:] + [axis])
+
+    inds = np.ndindex(a_view.shape[:-1])                                    #star
+    modes = jnp.empty(a_view.shape[:-1], dtype=a.dtype)
+    counts = jnp.empty(a_view.shape[:-1], dtype=jnp.int_)
+    for ind in inds:
+        modes[ind], counts[ind] = _mode1D(a_view[ind])
```

Comment:
JAX arrays are immutable, so you cannot assign their values this way.

But it looks like this is an attempt to map a 1D function to multiple dimensions... for performance reasons, we should lean toward doing this with `vmap` rather than by manual iteration.